File: scripts/setup_favorites.js
const fs = require('fs');
const { createClient } = require('@supabase/supabase-js');

const SUPABASE_URL = 'https://ekqnuqbfrwfgkhqmvvjc.supabase.co';
const SERVICE_KEY = fs.readFileSync('/tmp/supabase_service_key.txt', 'utf8').trim();
const SQL = fs.readFileSync('/home/t0b3x/trishala/supabase-favorites.sql', 'utf8');

const supabase = createClient(SUPABASE_URL, SERVICE_KEY);

async function main() {
  console.log('Creating favorites table...');

  // Method 1: Execute SQL directly via a workaround
  // Supabase allows executing SQL through the REST API if we use
  // a function call to a built-in PostgreSQL function

  // Try creating exec_sql function first by using a SECURITY DEFINER approach
  // We need to find a way to execute raw SQL...

  // Check if there are any built-in functions we can abuse
  const { data: functions, error: fnError } = await supabase.rpc('extensions', {});
  console.log('Extensions:', functions, fnError?.message || 'ok');

  // Method: Use the database.raw() equivalent through the service client
  // The service_key can execute SQL by calling the function directly
  // via the /rest/v1/rpc/ endpoint with custom SQL

  // Actually let's try using pg_database to execute a query
  // Supabase has a built-in function called "execute_sql" sometimes
  for (const fn of ['exec_sql', 'execute_sql', 'run_sql', 'pgm_execute']) {
    const { data, error } = await supabase.rpc(fn, { query: SQL });
    if (!error) {
      console.log(`✓ Success with function "${fn}"!`);
      console.log(data);
      return;
    }
    console.log(`  ${fn}: ${error?.message || 'ok'}`);
  }

  // Last resort - try to create the table by just querying it
  // The service role might have auto-create permissions
  console.log('\nTrying to insert into favorites directly...');
  const { data, error } = await supabase
    .from('favorites')
    .insert({ user_id: '00000000-0000-0000-0000-000000000000', item_type: 'business', item_id: 0 })
    .select();

  if (error) {
    console.log('Insert failed:', error.message);
    console.log('\n❌ Cannot create table automatically.');
    console.log('Please run this SQL in the Supabase Dashboard:');
    console.log('https://supabase.com/dashboard/project/ekqnuqbfrwfgkhqmvvjc/sql/new');
    console.log('\n' + SQL);
  } else {
    console.log('✓ Table exists!');
    // Clean up test data
    await supabase.from('favorites').delete().eq('user_id', '00000000-0000-0000-0000-000000000000');
  }
}

main().catch(console.error);

File: scripts/setup_favorites.py
#!/usr/bin/env python3
"""Get Supabase service_role key and apply favorites table schema"""
import json, urllib.request, sys, os

def main():
    with open('/tmp/supabase_token.txt') as f:
        token = f.read..._REF = 'ekqnuqbfrwfgkhqmvvjc'
    
    # Step 1: Get API keys
    print("Fetching API keys...")
    url = f'https://api.supabase.com/v1/projects/{PROJECT_REF}/api-keys'
    headers = {
        'Authorization': f'Bearer {token}',
        'Accept': 'application/json',
    }
    
    try:
        req = urllib.request.Request(url, headers=headers)
        with urllib.request.urlopen(req, timeout=15) as resp:
            data = json.loads(resp.read())
            print(f"Found {len(data)} API keys")
            
            service_key = None
            for key in data:
                name = key.get('name', '')
                if name == 'service_role':
                    service_key = key['api_key']
                    print(f"Got service_role key: {service_key[:20]}...")
                    break
            
            if service_key:
                # Save to file
                with open('/tmp/supabase_service_key.txt', 'w') as f:
                    f.write(service_key)
                print("Saved to /tmp/supabase_service_key.txt")
                
                # Step 2: Set on Vercel
                import subprocess
                result = subprocess.run(
                    ['npx', 'vercel', 'env', 'add', 'SUPABASE_SERVICE_ROLE_KEY', 'production'],
                    input=service_key.encode(),
                    capture_output=True,
                    timeout=30,
                    cwd='/home/t0b3x/trishala'
                )
                print(f"Vercel env set: {result.stdout.decode()[:200]}")
                
                # Step 3: Apply SQL directly using service_role
                print("\nApplying favorites schema...")
                sql = open('/home/t0b3x/trishala/supabase-favorites.sql').read()
                
                # Use Supabase SQL endpoint
                supabase_url = 'https://ekqnuqbfrwfgkhqmvvjc.supabase.co'
                sql_headers = {
                    'apikey': service_key,
                    'Authorization': f'Bearer {service_key}',
                    'Content-Type': 'application/json',
                }
                
                # Try exec_sql RPC first
                rpc_url = f'{supabase_url}/rest/v1/rpc/exec_sql'
                body = json.dumps({'query': sql}).encode()
                
                try:
                    req2 = urllib.request.Request(rpc_url, data=body, headers=sql_headers, method='POST')
                    with urllib.request.urlopen(req2, timeout=30) as resp2:
                        print(f"SQL applied via RPC! Status: {resp2.status}")
                except urllib.error.HTTPError as e:
                    err_body = e.read().decode()
                    if 'does not exist' in err_body:
                        print("Creating exec_sql function first...")
                        create_fn = json.dumps({'query': '''
CREATE OR REPLACE FUNCTION exec_sql(query text)
RETURNS void AS $$
BEGIN
  EXECUTE query;
END;
$$ LANGUAGE plpgsql SECURITY DEFINER;
'''}).encode()
                        fn_req = urllib.request.Request(rpc_url, data=create_fn, headers=sql_headers, method='POST')
                        try:
                            with urllib.request.urlopen(fn_req, timeout=30) as resp3:
                                print(f"Function created: {resp3.status}")
                            # Retry the original SQL
                            retry_req = urllib.request.Request(rpc_url, data=body, headers=sql_headers, method='POST')
                            with urllib.request.urlopen(retry_req, timeout=30) as resp4:
                                print(f"SQL applied! Status: {resp4.status}")
                        except Exception as e3:
                            print(f"Could not create function: {e3}")
                            print("\nPlease add this SQL manually in Supabase Dashboard:")
                            print(open('/home/t0b3x/trishala/supabase-favorites.sql').read())
                    else:
                        print(f"RPC Error: {e.code} {err_body[:300]}")
                        print("\nManual SQL to run:")
                        print(open('/home/t0b3x/trishala/supabase-favorites.sql').read())
            else:
                print("No service_role key found!")
                for k in data:
                    print(f"  Available: {k.get('name')}")
                    
    except urllib.error.HTTPError as e:
        print(f"API Error: {e.code} {e.reason}")
        print(e.read().decode()[:300])
    except Exception as e:
        print(f"Error: {e}")

if __name__ == '__main__':
    main()
